app/src/test/java/waiter/EchoServerTest.java

```diff
@@ -30,15 +31,28 @@ void tearDown() {
 
     @Test
     void catchesIOException() {
-        EchoServer echoServer = new EchoServer(new CommunicatorMock(new IOException()));
-        echoServer.start();
+        CommunicatorMock communicatorMock = new CommunicatorMock(true, new IOException());
+        EchoServer echoServer = new EchoServer(communicatorMock);
+        echoServer.start(4424);
         assertEquals("Sorry, an error occurred when sending/receiving a message, please try running the server and connecting again", output.toString().trim());
+        assertFalse(communicatorMock.communicateWasCalled);
     }
 
     @Test
     void catchesSocketException() {
-        EchoServer echoServer = new EchoServer(new CommunicatorMock(new SocketException()));
-        echoServer.start();
+        CommunicatorMock communicatorMock = new CommunicatorMock(true, new SocketException());
+        EchoServer echoServer = new EchoServer(communicatorMock);
+        echoServer.start(4424);
         assertEquals("Sorry, connection could not be established or has been broken, please try running the server and connecting again", output.toString().trim());
+        assertFalse(communicatorMock.communicateWasCalled);
+    }
+
+    // This is to cover lack of integration or E2E tests, could be replaced when these test type coverages are introduced
+    @Test
+    void callsCommunicate() {
+        CommunicatorMock communicatorMock = new CommunicatorMock(false, new IOException());
+        EchoServer echoServer = new EchoServer(communicatorMock);
+        echoServer.start(4425);
+        assertTrue(communicatorMock.communicateWasCalled);
```

Comment:
For context, Jon and I spoke about the lack of testing of this critical piece of logic. Despite the fact that it's testing a lot of the implementation, I think it's valuable to keep this test around. We might also want to look into adding e2e/integration tests.